### PR TITLE
#617 #626 fix stocktake actual quantity

### LINKED ISSUE #617: Stocktake crash when sorting by Counted column
Property 'countedTotalQuantity' does not exist on object type 'StocktakeItem' (on bugsnag)

### LINKED ISSUE #626: Stocktake value gets set to snapshot/lost
Build Number: 2.1.0-rc8

Description:
If a existing `Actual Quantity` value is tapped and highlighted, it'll be set to the snapshot quantity if you press next without changing the value to something different

Reproducible: yes

Reproduction Steps:
1. Have a stock take with a few items
2. Enter some values
3. If you select a value you have entered and press next, it'll be lost/set to what ever the snapshot value is.
4. You must enter a different value if you don't want it be set to snapshot quantity.

Comments:
We should test better. This was introduced in Dec and we only now picked it up. https://github.com/sussol/mobile/pull/580/files

### PR DESCRIPTION
**Pull request by Chris-Petty**
fixes #617 
fixes #626 

Fix not #412, tried to fix it using vscode but just totally messed up my commits as below...

So:

- Sorting by actual quantity no longer crashes
- Editing a actual quantity and setting it to previous actual quantity value doesn't set the countedTotalQuantity to snapshotQuantity
- Finalised stocktakes that have 0/no batches for countedTotalQuantity show 0 rather than blank.

### CODE REVIEW

File: src/database/DataTypes/StocktakeItem.js [hunk 1]
```diff
@@ -98,16 +98,6 @@ export class StocktakeItem extends Realm.Object {
...
-      // In case number is entered in Actual Quantity field, but it's the same as
-      // snapshot quantity, we want to remove 'Not Counted' placeholder
-      if (difference === 0) {
-        this.batches.forEach(stocktakeBatch => {
-          stocktakeBatch.countedTotalQuantity = stocktakeBatch.snapshotTotalQuantity;
-          database.save('StocktakeBatch', stocktakeBatch);
-        });
-        return;
-      }
-      // Actual Quantity is not snaphot quantity
```
Comment (Chris-Petty) [inline src/database/DataTypes/StocktakeItem.js]: I've tested this, the app correctly shows "Not Counted" when no value was set, not if the value is the same as snapshot.

If snapshot was 30, you counted 25. Then selected that field and entered 25 again, it'd see difference 0 and set it to 30.

File: src/pages/StocktakeEditPage.js [hunk 2]
```diff
@@ -101,6 +101,7 @@ export class StocktakeEditPage extends React.Component {
         sortDataType = 'string';
         break;
       case 'snapshotTotalQuantity':
+      case 'countedTotalQuantity':
```
Comment (Chris-Petty) [inline src/pages/StocktakeEditPage.js]: Fixes crash on sorting. Realm can't sort by getters.

File: src/pages/StocktakeEditPage.js [hunk 3]
```diff
@@ -115,12 +116,22 @@ export class StocktakeEditPage extends React.Component {
...
+      case 'countedTotalQuantity': {
+        let cellContents;
+        if (!stocktakeItem.hasCountedBatches) {
+          // If not editable (i.e. finalised) we want to show 0 rather than '' as
+          // placeholder isn't shown
+          cellContents = isEditable ? '' : 0;
+        } else {
+          cellContents = stocktakeItem.countedTotalQuantity;
+        }
+        // If finalised we want to show 0 rather than '' as placeholder isn't shown
```
Comment (Chris-Petty) [inline src/pages/StocktakeEditPage.js]: Fixes blank "ACTUAL QUANTITY" being blank when finalised line has no batches.

**Comment by Chris-Petty** [comment]
image

No content if you select the field and enter nothing and change focus (press next or tap nav bar)

File: src/pages/StocktakeEditPage.js [hunk 4]
```diff
@@ -115,12 +116,22 @@ export class StocktakeEditPage extends React.Component {
...
         return {
           type: isEditable ? 'editable' : 'text',
-          cellContents: stocktakeItem.hasCountedBatches ? stocktakeItem.countedTotalQuantity : '',
+          cellContents: cellContents,
```
Comment (andreievg) [inline src/pages/StocktakeEditPage.js]: Could just be cellContents,

File: src/pages/StocktakeEditPage.js [hunk 5]
```diff
@@ -115,12 +116,22 @@ export class StocktakeEditPage extends React.Component {
     switch (key) {
       default:
         return stocktakeItem[key];
-      case 'countedTotalQuantity':
+      case 'countedTotalQuantity': {
+        let cellContents;
+        if (!stocktakeItem.hasCountedBatches) {
+          // If not editable (i.e. finalised) we want to show 0 rather than '' as
+          // placeholder isn't shown
+          cellContents = isEditable ? '' : 0;
```
Comment (andreievg) [inline src/pages/StocktakeEditPage.js]: I am pretty sure if stocktakeItem does not have counted batches it gets removed during finalisation.

File: src/database/DataTypes/StocktakeItem.js (hunk 1 above)
Comment (andreievg) [reply src/database/DataTypes/StocktakeItem.js]: If there is more then one batch, and actual quantity is set to snapshot quantity we get this:
screen shot 2018-03-19 at 8 37 18 pm

File: src/pages/StocktakeEditPage.js (hunk 5 above)
Comment (Chris-Petty) [reply src/pages/StocktakeEditPage.js]: Nah the item remains annoyingly. If you expand it there are simply no rows. Maybe that's the bug I should be fixing!

Comment (andreievg) [reply src/pages/StocktakeEditPage.js]: I've tried numerous times (on master), If item has no counted quantity it get's removed from finalised stock take.  Can you please give a use case of where it remains please (on master).

File: src/database/DataTypes/StocktakeItem.js [hunk 6]
```diff
@@ -98,16 +98,6 @@ export class StocktakeItem extends Realm.Object {
       }
 
       let difference = quantity - this.countedTotalQuantity;
-      // In case number is entered in Actual Quantity field, but it's the same as
-      // snapshot quantity, we want to remove 'Not Counted' placeholder
-      if (difference === 0) {
-        this.batches.forEach(stocktakeBatch => {
-          stocktakeBatch.countedTotalQuantity = stocktakeBatch.snapshotTotalQuantity;
-          database.save('StocktakeBatch', stocktakeBatch);
```
Comment (andreievg) [inline src/database/DataTypes/StocktakeItem.js]: I've made a comment which you might have missed:

With this change, if there is more then one batch, and actual quantity is set to snapshot quantity we get this:
screen shot 2018-03-19 at 8 37 18 pm

File: src/database/DataTypes/StocktakeItem.js (hunk 1 above)
Comment (Chris-Petty) [reply src/database/DataTypes/StocktakeItem.js]: Ah I see! Will fix.

File: src/database/DataTypes/StocktakeItem.js (hunk 6 above)
Comment (Chris-Petty) [reply src/database/DataTypes/StocktakeItem.js]: Had missed, thanks!

File: src/pages/StocktakeEditPage.js (hunk 5 above)
Comment (Chris-Petty) [reply src/pages/StocktakeEditPage.js]: So here are some items getting stocktake.
image
Result:
image

So if the snapshot was 0, then it'll be included with no stocktakeItemBatches. If snapshot was more than 0 it works fine

Comment (andreievg) [reply src/pages/StocktakeEditPage.js]: So true. Nicely spotted

File: src/database/DataTypes/StocktakeItem.js (hunk 6 above)
Comment (andreievg) [reply src/database/DataTypes/StocktakeItem.js]: (changing back to countedTotalQuantity) @Chris-Petty, that was totally my bad. 
`if (difference === 0)` should ready:  `if (quantity === this.snapshotTotalQuantity)`

**Comment by andreievg** [comment]
@Chris-Petty, happy to merge if you reinstate:
```
if (quantity === this.snapshotTotalQuantity) {
        this.batches.forEach(stocktakeBatch => {
          stocktakeBatch.countedTotalQuantity = stocktakeBatch.snapshotTotalQuantity;
          database.save('StocktakeBatch', stocktakeBatch);
        });
        return;
      }
```
but with `if (quantity === this.snapshotTotalQuantity)`

File: src/pages/StocktakeEditPage.js (hunk 5 above)
Comment (andreievg) [reply src/pages/StocktakeEditPage.js]: I don't think it's a bug though, i have a distant recollection we wanted to keep them ?

Comment (Chris-Petty) [reply src/pages/StocktakeEditPage.js]: @andreievg I just spotted while sifting through code

(stocktake.js:adjustInventory)
```
// Delete all 'fresh' StocktakeBatches that have been created by stocktake
// but have not been changed
database.delete('StocktakeBatch', stocktakeBatches.filter(stocktakeBatch =>
  stocktakeBatch.snapshotTotalQuantity === 0 && stocktakeBatch.difference === 0));
```
So yea, it does check but doesn't touch the stocktakeItems, just clears out the batches.

**Comment by Chris-Petty** [comment]
I've put it back in @andreievg and seems to work as intended.

But I found a different unrelated problem hehe

Same as snapshot works correctly:
image

Reduced works correctly:
image

Increased works correctly:
image

But if you recount or correct a mistake, reducing the increased value
image

I think this anomaly is caused by the fact none of the 3 batches have a date. Giving them dates it works correctly.
image

**Comment by Chris-Petty** [comment]
I'm happy to make an issue for that so that this PR can move on in life.
#639

**Comment by andreievg** [comment]
@Chris-Petty thank you for all the testing!

**Pull request merged**